docs/src/pages/[...docPath].tsx

```diff
@@ -0,0 +1,120 @@
+import remarkDirective from "remark-directive";
+import { visit } from "unist-util-visit";
+import { h } from "hastscript";
+
+import {
+  generateFrontMatterTitleFromContent,
+  getMDPaths,
+  readMDFileFromPathOrIndex,
+  withIndexFile,
+  withInsertedCodeFromLinks,
+  withoutComments,
+} from "../model/md-generate";
+import Title from "../components/mdxComponents/Title";
+import Paragraph from "../components/mdxComponents/Paragraph";
+import CodeBlocks from "../components/mdxComponents/CodeBlocks";
+import Admonition from "../components/mdxComponents/Admonition";
+import DocumentationLayout from "../components/DocumentationLayout";
+
+const components = {
+  h2: Title.H2,
+  h3: Title.H3,
+  p: Paragraph,
+  code: CodeBlocks.Code,
+  pre: CodeBlocks.Pre,
+  tip: Admonition.Tip,
+  warning: Admonition.Warning,
+};
+
+/** @type {import('unified').Plugin<[], import('mdast').Root>} */
+function createCustomNodes() {
+  // @ts-ignore
+  return (tree) => {
+    visit(tree, (node) => {
+      if (
+        node.type === "textDirective" ||
+        node.type === "leafDirective" ||
+        node.type === "containerDirective"
+      ) {
+        // eslint-disable-next-line
+        const data = node.data || (node.data = {});
+        const hast = h(node.name, node.attributes);
+        // Create custom nodes from extended MD syntax. E.g. "tip"/"warning"
+        // @ts-ignore
+        data.hName = hast.tagName;
+        // @ts-ignore
+        data.hProperties = hast.properties;
+      }
+    });
+  };
+}
+
+interface IFrontMatter {
+  title: string;
+  description: string;
+  anchors?: string[];
+}
+interface IDocPage {
+  source: string;
+  frontMatter: IFrontMatter;
+}
+
+const DocPage: NextPage<IDocPage> = ({ source, frontMatter }): JSX.Element => {
+  return (
+    <DocumentationLayout
+      seo={{
+        title: frontMatter.title
+          ? frontMatter.title.concat(" | Hardhat")
+          : "Hardhat",
+        description:
+          frontMatter.description ||
+          "Ethereum development environment for professionals by Nomic Foundation",
+      }}
+    >
+      {/* @ts-ignore */}
+      <MDXRemote {...source} components={components} />;
+    </DocumentationLayout>
+  );
+};
+
+export default DocPage;
+
+export const getStaticProps: GetStaticProps = async (props) => {
+  const { params } = props;
+  // @ts-ignore
+  const fullName = withIndexFile(params.docPath, params.isIndex);
+  const source = readMDFileFromPathOrIndex(fullName);
+  const { content, data } = matter(source);
+  const formattedContent = withoutComments(withInsertedCodeFromLinks(content));
+
+  const mdxSource = await serialize(formattedContent, {
+    mdxOptions: {
+      remarkPlugins: [remarkDirective, createCustomNodes],
+      rehypePlugins: [],
+    },
+  });
+
+  return {
+    props: {
+      source: mdxSource,
+      frontMatter: {
+        title:
+          data.title ?? generateFrontMatterTitleFromContent(formattedContent),
```

Comment:
Could you please explain idea behind using coalescing here? e.g. passing empty string will reset a title? But what the usecase for that?

Reply:
So. Using this approach we can set title directly in yaml inside of the file. Otherwise we generate title from content

Reply:
I mean why you use `??` but not `||`. The difference is that it allows specifying empty string in frontmatter. Is this intentionally?